ludwig/trainers/trainer.py

```diff
@@ -659,6 +660,34 @@ def run_evaluation(
         # eval metrics on train
         self.eval_batch_size = max(self.eval_batch_size, progress_tracker.batch_size)
 
+        if self.dist_model.config_obj.model_type == "llm":
+            # Log the input and output of the LLM model
+            num_eval_examples = min(num_eval_examples, training_set.size)
+            inputs = {
+                i_feat.feature_name: torch.from_numpy(
+                    np.array(training_set.dataset[i_feat.proc_column][:num_eval_examples], copy=True)
+                ).to(self.device)
+                for i_feat in self.dist_model.input_features.values()
+            }
+            targets = {
+                o_feat.feature_name: torch.from_numpy(
+                    np.array(training_set.dataset[o_feat.proc_column][:num_eval_examples], copy=True)
+                ).to(self.device)
+                for o_feat in self.dist_model.output_features.values()
+            }
+            eval_example_outputs = self.dist_model((inputs, targets))["predictions"]
+            llm_outputs = {"inputs": [], "outputs": []}
+            input_list = list(inputs.values())[0]
+            for i in range(len(eval_example_outputs)):
+                clean_input = remove_left_padding(input_list[i], self.dist_model.tokenizer)[0]
```

Comment:
Instead of surfacing lower-level code and the raw forward pass, could we re-use `self.evaluation`, which could be modified to return predictions:

```python
    def evaluation(self, dataset, dataset_name, metrics_log, batch_size, progress_tracker):
        predictor = Predictor(
            self.dist_model,
            batch_size=batch_size,
            distributed=self.distributed,
            report_tqdm_to_ray=self.report_tqdm_to_ray,
            model=self.model,
        )
        ### Setting collect_predictions=True will return predictions.
        metrics, _ = predictor.batch_evaluation(dataset, collect_predictions=False, dataset_name=dataset_name)

        return append_metrics(self.model, dataset_name, metrics, metrics_log, progress_tracker)
```